### PR TITLE
[16.0][FIX] lighting: field invisibility based on parameters in lighting source lines

### PR DESCRIPTION
**Pull request by FrankC013**

### CODE REVIEW

**Review by eantones — CHANGES_REQUESTED** [review]
Please, make the changes we discussed in today's meeting

File: lighting/models/product_source_line.py [hunk 1]
```diff
@@ -380,6 +380,63 @@ def _compute_wattage_display(self):
...
+                    "led_chip_ids",
+                ],
+            }
+
+            key_fields = [
+                "is_led",
+                "is_integrated",
+                "is_lamp_included",
+                "is_max_wattage",
+            ]
```
Comment (eantones) [inline lighting/models/product_source_line.py]: Put this outside "for rec in self", maybe outside the class as a constant? What do you think?

**Review by eantones — CHANGES_REQUESTED** [review]

File: lighting/models/product_source_line.py [hunk 2]
```diff
@@ -20,6 +20,60 @@ class LightingProductSourceLine(models.Model):
     _rec_name = "type_id"
     _order = "sequence"
 
+    @api.model
```
Comment (eantones) [inline lighting/models/product_source_line.py]: Maybe using property will be similar to a constant.
```suggestion
    @property
```
chack for examples in Odoo

File: lighting/models/product_source_line.py [hunk 3]
```diff
@@ -20,6 +20,60 @@ class LightingProductSourceLine(models.Model):
...
+        """This method returns a list of key boolean fields that determine
+        the configuration state of a product."""
+        return [
+            "is_led",
+            "is_integrated",
+            "is_lamp_included",
+            "is_max_wattage",
+        ]
+
+    @api.model
```
Comment (eantones) [inline lighting/models/product_source_line.py]: Same

File: lighting/models/product_source_line.py [hunk 4]
```diff
@@ -20,6 +20,60 @@ class LightingProductSourceLine(models.Model):
...
+                "cri_min",
+                "color_consistency",
+                "special_spectrum_id",
+                "leds_m",
+                "led_chip_ids",
+            },
+        }
+
+    @api.model
+    def GROUP_MAP(self):
```
Comment (eantones) [inline lighting/models/product_source_line.py:60]: Same

File: lighting/models/product_source_line.py [hunk 5]
```diff
@@ -380,6 +440,64 @@ def _compute_wattage_display(self):
...
+
+    def update_dependent_fields(self, values):
+        self.ensure_one()
+        for field in self._get_irrelevant_dependent_fields(values):
+            # This is to be sure we don't update already falsy values
+            if not self[field]:
+                if field in values:
+                    del values[field]
+            else:
+                values[field] = None
```
Comment (eantones) [inline lighting/models/product_source_line.py]: We should have all these methods at the beginning just after the "constants" and comment the start and end of the block

File: lighting/models/product_source_line.py [hunk 6]
```diff
@@ -380,6 +440,64 @@ def _compute_wattage_display(self):
         string="Source",
     )
 
+    invisible_fields = fields.Json(
+        compute="_compute_invisible_fields",
+    )
+
+    def _get_all_dependent_fields(self):
```
Comment (eantones) [inline lighting/models/product_source_line.py]: To be like the others?
```suggestion
    def _get_dependent_all_fields(self):
```

File: lighting/models/product_source_line.py [hunk 7]
```diff
@@ -380,6 +440,64 @@ def _compute_wattage_display(self):
...
+        self.ensure_one()
+        all_fields = set()
+        for fields_g in self.FIELD_GROUP_MAP().values():
+            all_fields |= set(fields_g)
+        return all_fields
+
+    def _get_irrelevant_dependent_fields(self, values=None):
```
Comment (eantones) [inline lighting/models/product_source_line.py]: To be like the others?
```suggestion
    def  _get_dependant_irrelevant_fields(self, values=None):
```

File: lighting/models/product_source_line.py [hunk 8]
```diff
@@ -380,6 +440,64 @@ def _compute_wattage_display(self):
...
+        self.ensure_one()
+        return self._get_all_dependent_fields() - self._get_relevant_dependent_fields(
+            values=values
+        )
+
+    def _get_relevant_dependent_fields(self, values=None):
```
Comment (eantones) [inline lighting/models/product_source_line.py]: To be like the others?
```suggestion
    def  _get_dependant_relevant_fields(self, values=None):
```

File: lighting/models/product_source_line.py [hunk 9]
```diff
@@ -380,6 +440,64 @@ def _compute_wattage_display(self):
...
+            # This is to be sure we don't update already falsy values
+            if not self[field]:
+                if field in values:
+                    del values[field]
+            else:
+                values[field] = None
+
+    @api.depends("is_led", "is_integrated", "is_lamp_included", "is_max_wattage")
+    def _compute_invisible_fields(self):
+        # definition of VISIBLE fields (easier that the invisible)
```
Comment (eantones) [inline lighting/models/product_source_line.py]: Remove this

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — APPROVED** [review]

**Review by eantones — CHANGES_REQUESTED** [review]
CI is red

**Review by eantones — APPROVED** [review]

**Comment by FrankC013** [comment]
@eantones Continue WIP. One moment

**Comment by eantones** [comment]
I won't do anything til the combination of labels were the correct ones. If it's wor in progress i'll do nothing.

File: lighting/views/views.xml [hunk 10]
```diff
@@ -223,7 +224,10 @@
                                     widget="many2many_tags"
                                     options="{'color_field': 'color'}"
                                 />
-                                <field name="is_lamp_included" />
+                                <field
+                                    name="is_lamp_included"
+                                    attrs="{'column_invisible': [('parent.line_common_invisible_fields', 'in', ('is_lamp_included',))]}"
+                                />
```
Comment (eantones) [inline lighting/views/views.xml]: What would happen if a source has 3 lines and one line has 'lamp included' filled in but it is not a relevant field, and another line has 'lamp included' filled in and it is relevant? The column would then be displayed, and the column would show the 'is_lamp_included' activated for the line where it is not relevant, right?

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — APPROVED** [review]

**Pull request merged**